server/src/main/java/io/druid/server/coordination/SegmentChangeRequestHistory.java

```diff
@@ -0,0 +1,300 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.server.coordination;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.AbstractFuture;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import io.druid.java.util.common.IAE;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * This class keeps a bounded list of segment updates made on the server such as adding/dropping segments.
+ *
+ * Clients call addSegmentChangeRequest(DataSegmentChangeRequest) or addSegmentChangeRequests(List<DataSegmentChangeRequest>)
+ * to add segment updates.
+ *
+ * Clients call ListenableFuture<SegmentChangeRequestsSnapshot> getRequestsSince(final Counter counter) to get segment
+ * updates since given counter.
+ */
+public class SegmentChangeRequestHistory
+{
+  private static int MAX_SIZE = 1000;
+
+  private final int maxSize;
+
+  @GuardedBy("this")
+  private final List<Holder> changes;
+
+  @VisibleForTesting
+  @GuardedBy("waitingFutures")
+  final LinkedHashMap<CustomSettableFuture, Counter> waitingFutures;
+
+  private final ExecutorService singleThreadedExecutor;
+  private final Runnable resolveWaitingFuturesRunnable;
+
+  public SegmentChangeRequestHistory()
+  {
+    this(MAX_SIZE);
+  }
+
+  @VisibleForTesting
+  SegmentChangeRequestHistory(int maxSize)
+  {
+    this.maxSize = maxSize;
+    this.changes = new ArrayList<>(maxSize);
+
+    this.waitingFutures = new LinkedHashMap<>();
+
+    this.resolveWaitingFuturesRunnable = new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        resolveWaitingFutures();
+      }
+    };
+
+    this.singleThreadedExecutor = Executors.newSingleThreadExecutor(
+        new ThreadFactoryBuilder().setDaemon(true)
+                                  .setNameFormat(
+                                      "SegmentChangeRequestHistory"
+                                  )
+                                  .build()
+    );
+  }
+
+
+
+  /**
+   * Add batch of segment changes update.
+   */
+  public synchronized void addSegmentChangeRequests(List<DataSegmentChangeRequest> requests)
+  {
+    for (DataSegmentChangeRequest request : requests) {
+      if (changes.size() >= maxSize) {
+        changes.remove(0);
```

Comment:
This might be pre-mature optimization, but I worry about how this data structure is being used.  One option would be to view the list as a revolving buffer and essentially maintain an index of the oldest "valid" item.  You can replace and increment that on `add` and start your search from there wrapping back around.